docs/fetches-too-slow.md

```diff
@@ -25,11 +25,9 @@ Developer machines typically have network access most of the time, so Git object
 If a developer runs `git maintenance start` in their repository, this enables several background maintenance tasks, including [the `prefetch` task](https://git-scm.com/docs/git-maintenance#Documentation/git-maintenance.txt-prefetch).
 By default, this task fetches new objects from the origin server on an hourly basis, leading to the developer not waiting as long to get the latest refs during their foreground `git fetch` and `git pull` commands.
 
-### Improve directory structure
+### Shallow your directory structure
```

Comment:
I suggest dramatically scaling back or even eliminating this section. Sure, an overly deep directory structure will force Git to create more tree objects and traverse them when looking up data. But in practice, big directory trees (e.g., more than a few hundred entries), let alone giant ones (with thousands or tens of thousands of entries) are far more often problematic, and the solution to that problem often involves deepening the tree into more but smaller directories. For example, if you have

```
|- foo
|    |- file0000.txt
|    |- file0001.txt
|    |- file0002.txt
|    ⋮
|    |- file4999.txt
```

then every commit that makes any change to any of the contained files will require a new `foo` tree to be created. Since that tree has 5000 entries, it will be roughly 200 kB in size. That tree will delta well so won't take up much space on disk, but every time that it needs to be traversed, Git will have to un-delta it to its full 200 kB and parse it, both of which are expensive. A full repository traversal, for example when repacking objects, will have to expand and process every version of this directory. If, on the other hand, you shard this directory into

```
|- foo
|    |- dir00
|    |    |- file0000.txt
|    |    |- file0001.txt
|    |    |- file0002.txt
|    |    ⋮
|    |    |- file0099.txt
|    |- dir01
|    |    |- file0100.txt
|    |    ⋮
|    |    |- file0199.txt
|    ⋮
|    |- dir49
|         |- file4900.txt
|         ⋮
|         |- file4999.txt
```

, now a modification to a single file requires a 50-entry tree and a 100-entry tree to be created, which is much less new data than before (150 total entries vs 5000, plus the top-level directory). This will typically be very _good_ for Git performance.

So, of course don't gratuitously add unneeded nesting to your directory structure, but the main emphasis should be to be proactive about adding nesting if it helps reduce the greater evil of having very large directories.

Reply:
Thank you. Both of these sections are about "don't do pathological things", not common cases. I've not done a good job putting boundaries on when a workflow tips over into pathology. I'm aware of at least one GitHub customer who met the bar for pathologically-deep; from talking to you offline, you're aware of many who fall into pathologically-wide.

I suspect in practice, the case you're worried about happens _far_ more often. Let me see if I can address that - make the common thing more obvious - without totally eliminating the depth guidance.